docs/ocean/design_docs/vert-lagrang-remap/vert-lagrang-remap.rst

```diff
@@ -0,0 +1,540 @@
+
+The core of the proposed development is the ability to perform vertical 
+Lagrangian-remapping (VLR). When VLR is performed, the momentum, volume and 
+tracer conservation equations are first solved in a Lagrangian sense; that is, 
+layer thicknesses evolve such that there is no vertical transport across layer 
+interfaces. A target grid is then specified. For this stage of development, 
+we only require that the algorithm be able to use the existing grid 
+specifications. However, the implementation should be general enough that 
+a different grid could be used for remapping. This regridding step is followed 
+by a conservative remapping of velocity and scalars to the target grid. 
+
+Requirement: Grid is sufficiently conditioned for accuracy and stability
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/04
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+A severely deformed grid can lead to issues with the accuracy and stability of 
+the numerical methods used for the dynamics and thermodynamics. If layers are 
+allowed to become very thin they can also degrade the accuracy and stability of 
+the solution. Thus, a requirement of the VLR implementation is that there are 
+safeguards to prevent layers from becoming too thin or deformed. 
+
+Requirement: Ability to specify remapping method and its options
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/04
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+There are several choices to make for the remapping operation to balance accuracy, 
+stability and computational cost. To the extent that it is feasible, the user 
+should have the ability to control the order of accuracy of the remapping method,
+[Darren, other options here?] and other remapping options (e.g., via namelist options).
+
+Requirement: Support for existing time stepping schemes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/04
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+Vertical Lagrangian remapping should be supported (eventually) in both the RK4 
+and the split-explicit time stepping schemes currently implemented in MPAS-Ocean. 
+
+Requirement: not climate changing
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/15
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+For the same target grid, time-stepping scheme and problem, the solution for 
+layer thicknesses and key prognostic variables using VLR is within a non-
+climate-changing threshold value [presumably greater than machine precision?] 
+of the solution when VLR is not performed.
+
+Requirement: performance
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Date last modified: 2020/12/04
+
+Contributors: Darren Engwirda, Xylar Asay-Davis, Carolyn Begeman
+
+Performance is not degraded below an acceptable level *which is to be 
+determined*.
```

Comment:
@dengwirda @xylar I think we settled on leaving this TBD for now, but jump in if you have thoughts on this.

Reply:
Generally, any performance degradation should be kept to less than a few percent in overall throughput (SYPD). The smaller, the better - if we keep adding stuff, even small perf hits add up over time. Higher performance degradation can be justified for required changes that have significant (positive) impacts on simulation quality. We're happy to help with the design - catching things early is better so getting a performant design up front will avoid a complete refactor later.

Reply:
Thanks @philipwjones. When we get a little further along with knowing what needs to happen in the implementation phase we'll get in touch with you to hone the design for performance.

Reply:
I agree with @philipwjones, but I think we should revisit this on a second pass (after first getting the remap working, and ensuring consistency with the flux-form implementation).

With performance in mind, I can imagine rewriting some of the remap functionality from PPR direct in MPAS-O, to ensure compatibility with data-structures, eliminate copies to temp arrays, etc. Will be great to work with you on this @philipwjones, when we get a little further progressed.

Reply:
That's fine, but an early heads-up - for GPU performance, we want to see local tightly nested loops (with both horiz and vertical loops nested together). So please don't code these as 1-d column-based routines. We've been having to back out that design decision for a lot of other modules. See, for example:

https://acme-climate.atlassian.net/wiki/spaces/DOC/pages/1041499662/Fortran+GPU+Refactoring+Guide+Using+Directives

Reply:
@philipwjones Thanks for the link! That looks very useful.

Reply:
Totally agree @philipwjones! Perhaps we can arrange a meeting on this at some point early next year, and you can advise on best practice loop/data layouts, etc. The remap code has a somewhat different structure to "standard" advection --- I'd be happy to run through the formulation details too.